### ISSUE TITLE
Replace v3s16 and v3f with typedefs

### ISSUE THREAD

**Issue by nerzhul** [minetest/minetest#7037]
##### Issue type

 - Feature request
 
##### Minetest version

```
devel
```

##### OS / Hardware

Any

##### Summary

Name nodepos and worldpos in code using typedef to ensure consistency, maintenance and future migration (if required)

I suggest to replace nodepos (v3s16) with the typedef name nodepos_t and world position (v3f) with worldpos_t. It was done in network code, replacing u64 with session_t and permits to catch annoying type mismatches.

@minetest/engine are you okay with this ? Please vote on the post. I will open a PR asap as we have sufficient 👍 or 👎 

##### Steps to reproduce

None

**Comment by HybridDog** [comment]
`👍` You could also use doxygen to describe that e.g. nodepos_t represents the position of a node in the world and is implemented as 16 bit vector.

**Comment by Ekdohibs** [comment]
It might also make sense to replace `s16` with `coord_t` where sensical. Besides permitting to catch type mismatches, it is also a good first step to abstracting the world size limitation, which, even if it is kept, has no reason to be appearing in almost every single line of the minetest codebase.

**Comment by nerzhul** [comment]
@Ekdohibs yeah it's exact we can extend this to single coordinate too :), but you have nodecoord_t and worldcoord_t then, due to the double type

**Comment by HybridDog** [comment]
v3f represents floats (32 bit), doesn't it?

I've just tested the precision of 32 bit float near 30900 (representing mapblock boundaries):
`0, 0.00195312, -0.00195312`
```C
#include <stdio.h>

int main()
{
	float x = 30900.f;
	int *n = (int *)&x;
	int ns = *n - 1;
	int nb = *n + 1;
	float xs = *(float *)&ns;
	float xb = *(float *)&nb;
	printf("%g, %g, %g\n", *(float *)n - 30900.f, xb - 30900.f, xs - 30900.f);
	return 0;
}
```

**Comment by nerzhul** [comment]
v3f are world position (irrlicht map position), yes it's f32, float

**Comment by paramat** [comment]
>  to ensure consistency, maintenance

Hard to judge. Seems vague and low priority, can you explain the advantages more?
What do you mean by 'future migration'? Migration of what to what? However when deciding this we should not consider anything that might not happen.
>  it is also a good first step to abstracting the world size limitation

We should not consider a larger world size when deciding this, because that may not happen (even if it probably eventually will).
> permits to catch annoying type mismatches.

Are these really an issue? I work on these values a lot and can't remember these being an issue.

Seeing as we have limited dev time and there are more important things to do i suggest directing your efforts to more important things, like completing server-sent CSM, and either securing builtin against exploits or merging red-001's PR that makes that unnecessary.

Not having v3s16 and v3f makes code less clear and less readable, the type is not visible.
The advantages seem to be trivial issues about catching poor coding, how about we just code well instead.
For now 👎

**Comment by nerzhul** [comment]
i don't see why nodepos_t is less visible than v3s16 as you just need to know it's a v3s16.
do you know what is the C builtin size_t ?
for the CSM issues i have less time than boring PR like this as my brain time is limited these days on long periods

**Comment by paramat** [comment]
Completing CSM is much more important than this, are you abandoning server-sent CSM? This would mean all CSM will be removed, i hope this doesn't happen i want to see server-sent completed. But if you have no time for CSM i can't see it progressing at all.

**Comment by nerzhul** [comment]
missing time != abandon, and you cannot make a feature depend only on one developer, if it's the problem the whole software is broken

**Comment by paramat** [comment]
I'm just concerned because no other devs are working on CSM, they're busy with basic maintenance, i'll ask red-001 to work on server-sent. I would have thought server-sent CSM would interest you much more than boring type changes.

**Comment by nerzhul** [comment]
@paramat it's interesting, but i just don't get many time to work on it and find the good design, it's not a 2 line feature :)

**Comment by paramat** [comment]
Ok, but i'm a little cinfused, if you're lacking time and are interested in CSM and want to work on it why not work on CSM instead of unnecessary maintenance?

**Comment by red-001** [comment]
can we please stay on-topic here, CSM isn't related to renaming the types that are used for node positions.

**Comment by paramat** [comment]
It is on-topic, it's relevant to say 'i think our very limited time is best spent on more important things'. But anyway enough of that.

**Comment by nerzhul** [comment]
@red-001 is right, we are talking about code quality, not my lifetime

**Comment by paramat** [comment]
It's suggested that we replace all nodepos v3s16 with nodepos_t and define that in one place as v3s16.
The only advantage i can see is to make nodepos type easy to change, however changing that is a once-per-many-years thing (one of the types that is least likely to change) and is not decided yet. It's a huge decision to make and we can't assume it's going to happen, because that would be making the decision now. We should wait unltil after that decision is made, which is a long way off, probably MT 0.6.

I actually like the idea increasing world size, but i have concerns.
Does 'future migration' refer to this changing of co-ordinate types?

What other advantage is there? I can't see any.

If you mean it helps to prevent a coder accidently writing v3s32 or v2s16 for a nodepos, that is extremely rare, i can't remember the last time i saw such a mistake in a PR. By hiding the type it may actually make mistakes more common because it is now not clearly visible as a v3s16.

> i don't see why nodepos_t is less visible than v3s16 as you just need to know it's a v3s16.

This is what i mean, you now have to remember (as easy as it is do to so) it's type, it makes code less clear, less readable and working with code more effort. It's much more pleasant to have the type clearly visible. It will certainly irritate me.

So i don't see enough good reason to do this (yet).

**Comment by sofar** [comment]
> The only advantage i can see is to make nodepos type easy to change,

Patches like these increase robustness and make code maintenance easier, and the code becomes more consistent and we can rely better on things like type checking.

That in itself should be good enough reason to do this. That at some later point we could change coordinate systems is much more work than just this pach.

I consider patches like this part of the "we should have done this ages ago". This is something every developer should want, and if they don't understand why, then we need to do better education.

**Comment by kilbith** [comment]
> Seeing as we have **limited** dev time

Well, not for writing lengthy bullshits apparently... What about sharpening your coding skills, instead?

**Comment by paramat** [comment]
Ok, there are probably some good reasons i do not realise at the moment. I did politely ask for more details:
> Hard to judge. Seems vague and low priority, can you explain the advantages more?
What do you mean by 'future migration'? Migration of what to what?

But the author didn't answer. Sofar's answer was also a little vague (but appreciated thank you) but maybe it's too complex to go into detail over.
We have types throughout the code i can't see the harm in them, they make code more readable, if all types were abstracted code would be much less clear.
It's a good point that we shouldn't consider what might happen in future, world size increasing is desirable but controversial, problematic and not inevitable.

My comments about working on something else instead were unreasonable, sorry about that.
I was alarmed by what seemed to be a statement that nerzhul found CSM boring and not worth working on, that was probably a misunderstanding.

///////////////

kilbith, you're up to 6 or 7 out of 10 now, so are getting close to being banned from here. I actually don't want to see that happen as you are a good contributor. At any point i am willing to leave the past behind and get along with you if you decide to be civil.

**Comment by nerzhul** [comment]
@paramat having common typedef just make maintainance easier and code clearer about what type we ask about. It's just make dev and reviewer life easier, and securise against mistakes.
CSM is not boring, the boring part is to see complains and no constructive discussion about the solution :)

**Comment by Ekdohibs** [comment]
@paramat IMHO, it's precisely abstracting the types so that their name reflects their usage that makes the code clearer. Besides, whether world size increases or not, why should *every single line* of the code reflect that world size is limited to 16 bits? It has nothing to do with the vast majority of the code, and it should only appear where it matters.

**Comment by paramat** [comment]
nerzhul,
> having common typedef just make maintainance easier and code clearer about what type we ask about. It's just make dev and reviewer life easier, and securise against mistakes.

That is still vague but it may be somewhat true for some reasons i don't realise.
However conversely, consider that i work on mapgen and with these types more than anyone, and do not necessarily agree with any part of that statement.
I agree it will probably very very rarely avoid a typo, but i can't remember a co-ordinate type typo.
I think my experience counts for something.

Ekdohibs,
> whether world size increases or not, why should every single line of the code reflect that world size is limited to 16 bits? It has nothing to do with the vast majority of the code, and it should only appear where it matters.

I can't see any valid argument here, what harm is it? It's simply the type and the replacement is the also the same type. The type being seen or not seems a non-issue.
Clearer is debateable.

IRC http://irc.minetest.net/minetest-dev/2018-02-15#i_5231305
celeron55 wrote:
> if MT was still my personal project, i wouldn't do that (that's why it isn't so in the first place)

If i understand correctly it seems celeron55 would not do this.

Hmmmm was our mapgen dev for years and had a quality standard as high and probably higher than anyones, yet he never did this and i don't remember him suggesting it.
So doing this seems not as obvious as is suggested.
I'm also concerned about the suggestion to extend this to coord_t to abstract even more types. Please don't.

**Comment by nerzhul** [comment]
@paramat see just:

```
nodepos_t pos(1,2,3);
```

instead of

```
v3s16 pos(1,2,3);
```

**Comment by kilbith** [comment]
> kilbith, you're up to 6 or 7 out of 10 now

🖕 8 now.

**Comment by appgurueu** [comment]
I support this - it would allow easily changing world size.

**Comment by paramat** [comment]
Do you mean: Easy reducing of world size below the maximum (whatever that is)? If so, that can be done already with `mapgen_limit`.
Increasing maximum world size in MT is not easy, if that is what you meant.

**Comment by appgurueu** [comment]
I don't mean "easy reducing" - I mean increasing, and that could be at least made easier if @nerzhul makes a PR to use interfaces - I don't see any disadvantages to this.

**Comment by nerzhul** [comment]
it's also a way to ensure that the types on the objects are not mismatches when we develop. it's easy to use the wrong type by error. coord_t or nodepos_t has a real sense in terms of functional, whereas v3s16 it's just a storage of shorts.
If someone wants to take time to start to fix the types to a common name it's good and should not be forbidden. We are not managing dev budget on a company, it's a good improvement, yes with low priority, yes with low impact and no function change. But in terms of maintenance it's pretty good

**Comment by ClobberXD** [comment]
@paramat I'll try to help you understand the benefits of using typedefs here:

- If we ever were to ditch Irrlicht and use a different 3D engine, we'd have to go through the entire codebase and replace Irrlicht specific types with the new 3D engine's native types. It's easy to replace some of the lesser-used types, but types like v3f and v3s16 are virtually everywhere. If we use a type instead, we'd have to only change one line to make `nodepos_t` (for example) correspond to the new type. Of course, a lot more needs to be done when we actually have to switch to a new 3D engine (I just used it as an example), but code-wise, that's it. 

- Type mismatches, is another, very strong reason. Of course it's possible to code *very* carefully, but people generally tend to make mistakes (and it can get very hard to figure out the cause). A `v3s16` is just a "container" that can store 3 signed 16-bit integers. It's very easy to pass the wrong `v3s16` to a function by mistake, and then we'd potentially waste a lot of time trying to find the problem. But what if we made the function accept only arguments of type `nodepos_t`? Even if there are multiple typedefs corresponding to `v3s16`, the function can only accept `nodepos_t`. If we do pass the wrong variable, the code won't even compile, and we'd get a very straight-forward compile-time error that tells us what the problem is.

- Using typedefs (within reasonable limits) makes the code a whole lot more readable, as we can immediately make out the purpose of a variable just by looking at its type (of course, the function/variable name itself exists for this very purpose :P).

**Comment by TheTermos** [comment]
I'm not sure about this.

When I see v3s16 or v3f I immediately know they're 3d vectors and what types they consist of, that's what matters.
I can't think of any benefits of introducing high level logical context here, when I look at the type I don't care if it relates to world, a node or a duckie, that's what identifiers are for.

I'd expect this to make it less readable for me.
- they don't say anything useful about the type, you have to memorize what they represent
- it increases the overall number of types, as v3fs and v3s16s won't all disappear
- consistency, what about f32, f64, aabb3f, and hopefully upcoming usage of v3d and aabb3d? It introduces different conventions for related stuff.


Anyway, I'm putting up for consideration postponing working on this for after #9346 is resolved.
I can try if nobody else is willing. Did it once already, but the modification got too outdated in the meantime. I had to introduce v3d and aabb3d and overload type conversion operators between 32 and 64 versions, and anyway minetest code looks like mangled heap of guts and bones at places and is difficult to work with even without it being sugarcoated.

**Comment by sfan5** [comment]
> If we ever were to ditch Irrlicht and use a different 3D engine, we'd have to go through the entire codebase and replace Irrlicht specific types with the new 3D engine's native types

Not really no, you could use a typedef.

>  If we use a type instead, we'd have to only change one line to make nodepos_t (for example) correspond to the new type

Also no, if the new type uses lowercase member names (`x`, `y`, `z`), all code will be broken.
Even if that wasn't an issue, vectors are more than just a struct with three elements. The engine code uses methods like `getLength`, `getLengthSQ`, `getDistanceFrom` or `normalize`, these have to exist in a potential new type too or code will not compile.

> Using typedefs (within reasonable limits) makes the code a whole lot more readable,

`nodepos_t` is a lot longer than `v3s16` and you can't even infer its meaning from the name.

**Comment by ClobberXD** [comment]
> Not really no, you could use a typedef.

Yes, exactly. Unless you mean typedef'ing Irrlicht types?

> Even if that wasn't an issue, vectors are more than just a struct with three elements. The engine code uses methods like getLength, getLengthSQ, getDistanceFrom or normalize, these have to exist in a potential new type too or code will not compile.

Agreed.

> nodepos_t is a lot longer than v3s16 and you can't even infer its meaning from the name.

Longer, yes, but more intuitive, IMHO. `nodepos_t` is (or rather, would be) used for node positions, I assume, and the typedef is more descriptive than `v3s16`.

Typedefs have the unfortunate consequence of obscuring the actual type, but any standard IDE can reveal the actual type on cursor hover or something (if not, there's `grep`).

**Comment by sfan5** [comment]
> Unless you mean typedef'ing Irrlicht types?

Yes, that's what I mean.
If we ever switch to a new engine, existing types such as `v3s16` will either be aliased to the respective vector type of the new engine or a reimplementation of them will be kept and they can be converted to the engine type where necessary.

Turning `v3s16` into `nodepos_t` doesn't help with switching to a new engine, it only helps if you want to extend the coordinates to e.g. 32-bit in the future.

**Comment by ClobberXD** [comment]
Interesting, didn't think of that possibility.

**Comment by nerzhul** [comment]
v3f can be pos_t and v3s16 can be either nodepos_t or npos_t
